clients/include/argument_model.hpp

```diff
@@ -0,0 +1,152 @@
+        {
+            return
+                [](auto&& func, const Arguments& arg, auto T)
+                {
+                    func("alpha", arg.get_alpha<decltype(T)>());
+                };
+        }
+    };
+
+    // Specialization for e_beta
+    template <>
+    struct apply<e_beta>
+    {
+        auto operator()()
+        {
+            return
+                [](auto&& func, const Arguments& arg, auto T)
+                {
+                    func("beta", arg.get_beta<decltype(T)>());
+                };
+        }
+    };
+    // clang-format on
+
+public:
+    void log_perf(rocblas_ostream& name_line,
+                  rocblas_ostream& val_line,
+                  const Arguments& arg,
+                  double           gpu_us,
+                  double           gflops,
+                  double           gpu_bytes,
+                  double           cpu_us,
+                  double           norm1,
+                  double           norm2)
+    {
+        constexpr bool has_batch_count = has(e_batch_count);
+        rocblas_int    batch_count     = has_batch_count ? arg.batch_count : 1;
+        rocblas_int    hot_calls       = arg.iters < 1 ? 1 : arg.iters;
+
+        // per/us to per/sec *10^6
+        double rocblas_gflops = gflops * batch_count * hot_calls / gpu_us * 1e6;
+        double cblas_gflops   = gflops * batch_count / cpu_us * 1e6;
+
+        // bytes/us to GB/s = 10^6 * 10^-9 = 10^-3
+        double rocblas_GBps = gpu_bytes * batch_count / gpu_us / 1e3;
+
+        name_line << "rocblas-Gflops,rocblas-GB/s,rocblas-us,";
+        val_line << rocblas_gflops << "," << rocblas_GBps << "," << gpu_us << ",";
+
+        if(arg.unit_check || arg.norm_check)
+        {
+            name_line << "CPU-Gflops,CPU-us,";
+            val_line << cblas_gflops << "," << cpu_us << ",";
+            if(arg.norm_check)
+            {
+                name_line << "norm_error_host_ptr,norm_error_device_ptr,";
+                val_line << norm1 << "," << norm2 << ",";
+            }
+        }
+    }
+
+    template <typename T>
+    void log_args(rocblas_ostream& str,
+                  const Arguments& arg,
+                  double           gpu_us,
+                  double           gflops,
+                  double           gpu_bytes = 0,
+                  double           cpu_us    = 0,
+                  double           norm1     = 0,
+                  double           norm2     = 0)
+    {
+        rocblas_ostream name_list;
+        rocblas_ostream value_list;
+
+        // Output (name, value) pairs to name_list and value_list
+        auto print = [&, delim = ""](const char* name, auto&& value) mutable {
+            name_list << delim << name;
+            value_list << delim << value;
+            delim = ",";
+        };
+
+        // Apply the arguments, calling print on each one
+        // TODO: Replace with C++17 fold comma expression
+        (void)(int[]){(apply<Args>{}()(print, arg, T{}), 0)...};
```

Comment:
This should be put into a separate function to apply a function to each args as there is a lot of sublte C++ quirks happening in a small amount of code. Ideally, something like this:

```cpp
struct swallow
{
    template <class... Ts>
    constexpr swallow(Ts&&...)
    {
    }
};
template <class F, class... Ts>
constexpr void each_args(F f, Ts&&... xs)
{
    swallow{(f(std::forward<Ts>(xs)), 0)...};
}
```

And then this can be written as:

```cpp
each_args([&](auto&& x) {
    apply<Args>{}()(print, x, T{});
}, xs...);
```

Reply:
I think it would make the code harder to understand if we added additional functions/classes/lambdas. The `apply` function, as it stands now, does what we need it to for a specific purpose.

I'd prefer to use `std::apply`, but we're using `-std=c++14` now. I wonder if we should make the leap to `-std=c++17` yet. With C++17 we'd gain:
- `if constexpr`
- fold expressions
- `std::apply`
- `std::variant`
- automatic deduction of class template arguments from constructor arguments
- ...

Reply:
It has been moved into an `ArgumentsHelper` class. I don't know why I thought it depended on the template parameters. It's much simpler now, since it uses `constexpr` lambda variable templates.